_posts/2020-09-06-bench-rust-kotlin-microservices.md

```diff
@@ -0,0 +1,448 @@
+
+Further right in the graph, Kotlin implementations quickly saturate the allocated CPU portion,
+Rust following ~2 steps later.
+
+I'm curious that Http4k manages to consume a bit more CPU than the 15 CPU-seconds assigned by Docker.
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/cpu_per_request.svg" />
+
+Consumed CPU time divided by successful requests per second.
+
+I'm not sure how to explain the improving efficiency of Rust (Actix) with the increasing number of concurrent connections.
+Some constant background work, that is however not happening when idle?
+Improved efficiency of a batch I/O?
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/cpu_vs_requests.svg" />
+
+My favourite metric at the very end;
+the running total of successfully served requests (time-integrated throughput)
+over the cumulative sum of used CPU time, or *bang for the buck* for short.
+Going up means serving requests, while going right means consuming CPU.
+Line slope corresponds to CPU efficiency.
+
+If we take the furthest point of each implementation
+and divide the total processor time consumed by the total served request count,
+we arrive at *weighted average CPU time per request*,
+which incorporates CPU time needed to start the service and get it up to speed.
+
+If we allow ourselves a blatant simplification,
+we can transform the number to more comprehensible cost per billion requests,
+assuming an example price $0.0275 per vCPU⋅hour and unrealistic perfect resource utilization.
+
+|                           | Http4k      | Ktor        | Actix       |
+| ------------------------- | ----------: | ----------: | ----------: |
+| CPU time per request      | 560&nbsp;µs | 460&nbsp;µs | 170&nbsp;µs |
+| Cost per billion requests |        $4.3 |        $3.5 |        $1.3 |
+
+Please take these numbers with a grain of salt.
+Because benchmarked instances run only for a couple of minutes,
+we give excessive weight to startup costs.
+This effect may be only partially compensated by the fact
+that real microservices would probably run under a load much below their saturation points,
+thus taking much longer to serve an equal number of requests.
+
+## Conclusion
+
+In this post, I wanted to go beyond usual benchmarks by having an extended set of metrics
+to show more nuances of tested technologies
+and discuss some of their implications.
+I desired to compare each framework in its best shape,
+so I ended up doing 3 qualification benchmarks,
+which may hopefully serve people also on their own.
+All implementations, tools and data are open-sourced,
+so the benchmarks should be in theory reproducible by anybody;
+also it should be easy to add new implementations to the mix.
+
+The post was only about runtime characteristics,
+which is just one of many aspects of each technology.
+Comparing developer experience would be material for a whole new write-up!
```

Comment:
@goodhoko want to participate on that? :)

Reply:
Had I been on dev.to I could write a post comparing 3 different frameworks based on my deep experience of implementing a single endpoint in them. However, that's something I'd probably wouldn't want to present myself with. .))


Otoh I could invest my time and finish the ktor and vertx services to be feature complete and **then** write the post. But you know...
https://www.youtube.com/watch?v=HL1UzIK-flA&ab_channel=RihannaVEVO

Reply:
Ack. Let's see if there is some feedback requesting such a post, the hint is pretty obvious.